cmd/clusterctl/pkg/client/repository/components.go

```diff
@@ -64,3 +74,409 @@ type Components interface {
+	// the default target namespace is the namespace object defined in the component yaml read from the repository, if any
+	defaultTargetNamespace, err := inspectTargetNamespace(objs)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to detect default target namespace")
+	}
+
+	// Ensures all the provider components are deployed in the target namespace (apply only to namespaced objects)
+	// if targetNamespace is not specified, then defaultTargetNamespace is used. In case both targetNamespace and defaultTargetNamespace
+	// are empty, an error is returned
+
+	if targetNamespace == "" {
+		targetNamespace = defaultTargetNamespace
+	}
+
+	if targetNamespace == "" {
+		return nil, errors.New("target namespace can't be defaulted. Please specify a target namespace")
+	}
+
+	objs = fixTargetNamespace(objs, targetNamespace)
+
+	// ensure all the ClusterRoleBinding which are referencing namespaced objects have the name prefixed with the namespace name
+	objs, err = fixClusterRoleBindings(objs, targetNamespace)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to fix ClusterRoleBinding names")
+	}
+
+	// inspect the list of objects for the default watching namespace
+	// the default watching namespace is the namespace the controller is set for watching in the component yaml read from the repository, if any
+	defaultWatchingNamespace, err := inspectWatchNamespace(objs)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to detect default watching namespace")
+	}
+
+	// if the requested watchingNamespace is different from the defaultWatchingNamespace, fix it
+	if defaultWatchingNamespace != watchingNamespace {
+		objs, err = fixWatchNamespace(objs, watchingNamespace)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to set watching namespace")
+		}
+	}
+
+	objs = addLabels(objs, provider.Name())
+
+	return &components{
+		Provider:          provider,
+		version:           version,
+		variables:         variables,
+		targetNamespace:   targetNamespace,
+		watchingNamespace: watchingNamespace,
+		objs:              objs,
+	}, nil
+}
+
+var variableRegEx = regexp.MustCompile(`\${\s*([A-Z0-9_]+)\s*}`)
+
+func inspectVariables(data []byte) []string {
+	variables := map[string]string{}
+	match := variableRegEx.FindAllStringSubmatch(string(data), -1)
+
+	for _, m := range match {
+		submatch := m[1]
+		if _, ok := variables[submatch]; !ok {
+			variables[submatch] = ""
+		}
+	}
+
+	var ret []string // nolint
+	for v := range variables {
+		ret = append(ret, v)
+	}
+
+	sort.Strings(ret)
+
+	return ret
+}
+
+func replaceVariables(yaml []byte, variables []string, configVariablesClient config.VariablesClient) ([]byte, error) {
+	tmp := string(yaml)
+	var missingVariables []string
+	for _, key := range variables {
+		val, err := configVariablesClient.Get(key)
+		if err != nil {
+			missingVariables = append(missingVariables, key)
+			continue
+		}
+		exp := regexp.MustCompile(`\$\{\s*` + key + `\s*\}`)
```

Comment:
Maybe let's make this a different function so we can test it separately?

Reply:
This function is a simple a for loop around a replace and it is already tested. 
I don't think we will get a big improvement in splitting out the replace, but I will be happy to add more test if necessary

Reply:
Yeah, I was more talking about the regex generation